src/libs/TransactionUtils.ts

```diff
@@ -0,0 +1,359 @@
+import Onyx, {OnyxCollection} from 'react-native-onyx';
+import {format, parseISO, isValid} from 'date-fns';
+import CONST from '../CONST';
+import ONYXKEYS from '../ONYXKEYS';
+import DateUtils from './DateUtils';
+import * as NumberUtils from './NumberUtils';
+import {RecentWaypoint, ReportAction, Transaction} from '../types/onyx';
+import {Receipt, Comment, WaypointCollection} from '../types/onyx/Transaction';
+
+type AdditionalTransactionChanges = {comment?: string};
+
+type TransactionChanges = Partial<Transaction> & AdditionalTransactionChanges;
+
+let allTransactions: OnyxCollection<Transaction> = {};
+
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.TRANSACTION,
+    waitForCollectionCallback: true,
+    callback: (value) => {
+        if (!value) {
+            return;
+        }
+        allTransactions = Object.fromEntries(Object.entries(value).filter(([, transaction]) => !!transaction));
+    },
+});
+
+/**
+ * Optimistically generate a transaction.
+ *
+ * @param amount – in cents
+ * @param [existingTransactionID] When creating a distance request, an empty transaction has already been created with a transactionID. In that case, the transaction here needs to have it's transactionID match what was already generated.
+ */
+function buildOptimisticTransaction(
+    amount: number,
+    currency: string,
+    reportID: string,
+    comment = '',
+    created = '',
+    source = '',
+    originalTransactionID = '',
+    merchant = '',
+    receipt: Receipt = {},
+    filename = '',
+    existingTransactionID: string | null = null,
+    category = '',
+    tag = '',
+    billable = false,
+): Transaction {
+    // transactionIDs are random, positive, 64-bit numeric strings.
+    // Because JS can only handle 53-bit numbers, transactionIDs are strings in the front-end (just like reportActionID)
+    const transactionID = existingTransactionID ?? NumberUtils.rand64();
+
+    const commentJSON: Comment = {comment};
+    if (source) {
+        commentJSON.source = source;
+    }
+    if (originalTransactionID) {
+        commentJSON.originalTransactionID = originalTransactionID;
+    }
+
+    // For the SmartScan to run successfully, we need to pass the merchant field empty to the API
+    const defaultMerchant = !receipt || Object.keys(receipt).length === 0 ? CONST.TRANSACTION.DEFAULT_MERCHANT : '';
+
+    return {
+        transactionID,
+        amount,
+        currency,
+        reportID,
+        comment: commentJSON,
+        merchant: merchant || defaultMerchant,
+        created: created || DateUtils.getDBTime(),
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+        receipt,
+        filename,
+        category,
+        tag,
+        billable,
+    };
+}
+
+function hasReceipt(transaction: Transaction | undefined | null): boolean {
```

Comment:
```suggestion
function hasReceipt(transaction?: Transaction | null): boolean {
```

Reply:
@blazejkustra This way, it will be possible to call `TransactionUtils.hasReceipt();` without any arguments, I am not sure that it's expected